Use Schema.get_attribute rather than utils.get_value in Relationship

scmmmh opened the pull request:
When using include_resource_linkage in relationships the code used utils.get_value to get the "id" attribute's value. However, I believe that it should use the link Schema's get_attribute method to enable schemas to override how the attribute is accessed (as is the case for all scalar attributes).

If the Schema's get_attribute has not been overridden, then this will change nothing in the code's behaviour. If it has been overridden, then the behaviour is now consistent.

Works for both marshmallow < 3 and >=3, but includes an at-runtime check for the marshmallow version. Not the most elegant solution, but I am stuck for anything else.

As far as I can see would fix #53.

scmmmh committed:
Use Schema.get_attribute rather than utils.get_value in Relationship

scmmmh commented:
Ignore for the moment, while I check the failing tests.

scmmmh committed:
Fix Schema.get_attribute parameter order for marshmallow 2.13. Add schema to unit tests.

scmmmh commented:
Fixed the issues (I was running with the dev marschmallow 3).
This will now work with marshmallow 2.13, but obviously fails for 3.
Happy for any guidance to make it pass in both scenarios.

scmmmh committed:
Make it compatible with marshmallow >=3 and < 3

scmmmh commented:
Now passes for both conditions, but includes an at-runtime check for the marshmallow version. Not the most elegant solution, but I am stuck for anything else.

scmmmh committed:
Missed a debug statement.

sloria commented on tests/test_fields.py:
```diff
@@ -40,7 +40,7 @@ def test_include_resource_linkage_single(self, post):
             related_url='/posts/{post_id}/author/',
             related_url_kwargs={'post_id': '<id>'},
-            include_resource_linkage=True, type_='people'
+            include_resource_linkage=True, type_='people', schema='PostSchema'
```
This indicates a breaking change. Before this PR, `schema` was only a necessary argument if passing `include_data=True`. Could you add a check to see if a schema is available before trying to use its `get_attribute` method?

sloria commented on marshmallow_jsonapi/fields.py:
```diff
@@ -134,12 +134,16 @@ def get_resource_linkage(self, value):
             resource_object = {
                 'type': self.type_,
-                'id': _stringify(get_value(value, self.id_field, value))
+                'id': _stringify(self.schema.get_attribute(value, self.id_field, value)
```
There's enough duplicated logic here that it might be worth separating this into a private helper (e.g. `_get_id(object)`).

sloria reviewed (changes requested):
Thanks @scmmmh for the PR! My main concern is the backwards-incompatibility, but that can be addressed. Could you also add a new test for this behavior?

Once those are done, this is good to merge.

scmmmh committed:
Handle missing schemas for backwards compatability. Added unit tests.

scmmmh commented:
I had misremembered that as the schema also being required for include_resource_linkage.

I have added a _get_id method to the Relationship. I just wonder whether this is better factored out as a more generic schema_get_value function into utils, which would allow for a dynamic definition of the method, as with the base get_value function?

I have added three test cases so that it is tested both with and without the schema attribute.

sloria commented on marshmallow_jsonapi/fields.py:
```diff
@@ -215,6 +215,18 @@ def _serialize_included(self, value):
+        if _MARSHMALLOW_VERSION_INFO[0] >= 3:
+            try:
+                return self.schema.get_attribute(value, self.id_field, value)
+            except ValueError:
```
I realize this handles the `ValueError` raised by `self.schema`, but there's a risk we might be swallowing a `ValueError`  raised in `get_attribute`. Therefore, I think LBYL is more appropriate in this case:

```python
if self.__schema:
    return self.schema.get_attribute(...)
```

sloria reviewed (changes requested):
Thanks for making those changes. Factoring out the code to a utility function isn't a bad idea, but I think it's unnecessary indirection for now; after all, it will only be used in one place.

Just one suggested change, and I think this is good to go.

scmmmh committed:
Explicitly check for the schema in _get_id

scmmmh commented:
I had not considered that. I have made the change.

sloria commented on tests/test_fields.py:
```diff
@@ -47,6 +47,17 @@ def test_include_resource_linkage_single(self, post):
+        field = Relationship(
+            related_url='/posts/{post_id}/author/',
+            related_url_kwargs={'post_id': '<id>'},
+            include_resource_linkage=True, type_='people', schema='PostSchema'
```
Does this test fail without your change? `PostSchema` does not override `get_attribute`.

scmmmh replied:
I do not believe it should, as the base Schema provides a get_attribute implementation which simply uses the utils.get_value function. Do you want me to add a test case with schema that overrides get_attribute?

sloria replied:
@scmmmh Yes, we need tests that would fail without your new code

scmmmh replied:
Will do.

scmmmh committed:
Merge https://github.com/marshmallow-code/marshmallow-jsonapi into dev

scmmmh committed:
Added a test case that fails with the old _get_id code

scmmmh committed:
Handle both pre 3 and 3 of marshmallow

scmmmh committed:
Fixed marshmallow version bug

scmmmh committed:
Revert the last change

scmmmh committed:
Finally fix version bug
